File: utils/validation.js
export class BlueprintValidationError extends Error {
    constructor(message, field = null) {
        super(message);
        this.name = 'BlueprintValidationError';
        this.field = field;
    }
}

export function validateBlueprint(blueprint) {
    const errors = [];

    if (!blueprint) {
        throw new BlueprintValidationError('Blueprint is null or undefined');
    }

    if (typeof blueprint !== 'object') {
        throw new BlueprintValidationError(`Blueprint must be an object, got ${typeof blueprint}`);
    }

    if (!blueprint.global_aesthetic || typeof blueprint.global_aesthetic !== 'object') {
        throw new BlueprintValidationError('Missing or invalid global_aesthetic object', 'global_aesthetic');
    }

    const aesthetic = blueprint.global_aesthetic;
    const requiredAestheticFields = ['art_style', 'color_grading', 'lighting_setup'];

    for (const field of requiredAestheticFields) {
        if (!aesthetic[field] || typeof aesthetic[field] !== 'string') {
            errors.push(`global_aesthetic.${field} must be a non-empty string`);
        }
    }

    if (!Array.isArray(blueprint.chronological_shots)) {
        throw new BlueprintValidationError('Missing or invalid chronological_shots array', 'chronological_shots');
    }

    if (blueprint.chronological_shots.length === 0) {
        throw new BlueprintValidationError('chronological_shots cannot be empty', 'chronological_shots');
    }

    const requiredShotFields = [
        'shot_index', 'duration_seconds', 'camera_direction',
        'framing_type', 'action_and_motion', 'environment_context', 'negative_elements'
    ];

    for (let i = 0; i < blueprint.chronological_shots.length; i++) {
        const shot = blueprint.chronological_shots[i];

        if (!shot || typeof shot !== 'object') {
            errors.push(`Shot ${i}: must be an object`);
            continue;
        }

        for (const field of requiredShotFields) {
            if (shot[field] === undefined || shot[field] === null) {
                errors.push(`Shot ${i}: missing required field "${field}"`);
            }
        }

        if (typeof shot.shot_index !== 'number' || shot.shot_index < 0) {
            errors.push(`Shot ${i}: shot_index must be a non-negative number`);
        }

        if (typeof shot.duration_seconds !== 'number' || shot.duration_seconds <= 0) {
            errors.push(`Shot ${i}: duration_seconds must be a positive number`);
        }

        for (const field of ['camera_direction', 'framing_type', 'action_and_motion', 'environment_context']) {
            if (typeof shot[field] !== 'string') {
                errors.push(`Shot ${i}: ${field} must be a string`);
            }
        }

        if (!Array.isArray(shot.negative_elements)) {
            errors.push(`Shot ${i}: negative_elements must be an array`);
        }
    }

    if (errors.length > 0) {
        throw new BlueprintValidationError(`Validation failed:\n  - ${errors.join('\n  - ')}`);
    }

    return true;
}

export function validateVideoMetadata(metadata) {
    if (!metadata) return false;

    const required = ['filename', 'duration_seconds', 'width', 'height'];
    for (const field of required) {
        if (metadata[field] === undefined || metadata[field] === null) {
            return false;
        }
    }

    return metadata.duration_seconds > 0 && metadata.width > 0 && metadata.height > 0;
}

export function sanitizeBlueprint(blueprint) {
    if (!blueprint) return null;

    const sanitized = {
        global_aesthetic: {
            art_style: blueprint.global_aesthetic?.art_style || 'unknown',
            color_grading: blueprint.global_aesthetic?.color_grading || 'unknown',
            lighting_setup: blueprint.global_aesthetic?.lighting_setup || 'unknown',
        },
        chronological_shots: []
    };

    for (const shot of (blueprint.chronological_shots || [])) {
        sanitized.chronological_shots.push({
            shot_index: typeof shot.shot_index === 'number' ? shot.shot_index : 0,
            duration_seconds: typeof shot.duration_seconds === 'number' && shot.duration_seconds > 0 ? shot.duration_seconds : 5,
            camera_direction: shot.camera_direction || 'static camera',
            framing_type: shot.framing_type || 'medium shot',
            action_and_motion: shot.action_and_motion || 'no action',
            environment_context: shot.environment_context || 'unknown environment',
            negative_elements: Array.isArray(shot.negative_elements) ? shot.negative_elements : [],
        });
    }

    return sanitized;
}

File: utils/validation.py
from __future__ import annotations

from typing import Any

from pydantic import ValidationError

from src.schemas.blueprint import UniversalBlueprint


def validate_video_metadata(metadata: dict[str, Any] | None) -> bool:
    if not metadata:
        return False

    required = ["filename", "duration_seconds", "width", "height"]
    for field in required:
        if metadata.get(field) is None:
            return False

    return bool(metadata["duration_seconds"] > 0 and metadata["width"] > 0 and metadata["height"] > 0)


class BlueprintValidationError(Exception):
    def __init__(self, message: str, field: str | None = None) -> None:
        super().__init__(message)
        self.name = "BlueprintValidationError"
        self.field = field


def validate_blueprint(blueprint: dict[str, Any]) -> bool:
    try:
        UniversalBlueprint(**blueprint)
        return True
    except ValidationError as e:
        errors = []
        for error in e.errors():
            loc = ".".join(str(loc_part) for loc_part in error["loc"])
            msg = error["msg"]
            errors.append(f"{loc}: {msg}")
        raise BlueprintValidationError("Validation failed:\n  - " + "\n  - ".join(errors)) from e


def _coerce_shot(shot: dict[str, Any], index: int) -> dict[str, Any]:
    out = dict(shot)
    if "shot_index" not in out or out["shot_index"] is None:
        out["shot_index"] = index
    ne = out.get("negative_elements")
    if isinstance(ne, str):
        out["negative_elements"] = [ne]
    elif ne is None:
        out["negative_elements"] = []
    if out.get("frame_references") is None:
        out["frame_references"] = []
    return out


def sanitize_blueprint(blueprint: dict[str, Any] | None) -> dict[str, Any] | None:
    if not blueprint or not isinstance(blueprint, dict):
        return None
    try:
        data = dict(blueprint)
        shots = data.get("chronological_shots", [])
        if isinstance(shots, list):
            data["chronological_shots"] = [_coerce_shot(s, i) for i, s in enumerate(shots)]
        blueprint_obj = UniversalBlueprint(**data)
        return blueprint_obj.model_dump()
    except (ValidationError, TypeError):
        return None
